DMTN-148.tex

```diff
@@ -50,9 +50,276 @@
+use the Butler collection that the calibration was generated in (which
+will enforce uniqueness on simultaneous processing).  Calibrations
+that are ingested from git could use the repository SHA1 hash, which
+would provide similar uniqueness.
+
+Although not formally a ``classic'' calibration product, the \verb|SKY|
+calibration is produced, stored, and accessed the same way.
+
+\subsection{``User-curated'' Calibrations}
+
+The other inputs necessary for ISR have generally been described as
+"user-curated".  These products are either defined by the \verb|CAMERA| as
+part of the afw.cameraGeom description of the instrument, or are
+variously persisted and accessed in unique ways.  Describing a path to
+fixing this is a goal of this tech note.
+
+As part of the \verb|DEFECTS| rework (RFC-595), there is now a fixed and
+generic way to handle the ingest, if not necessarily the construction,
+of that product.  The \verb|QE_CURVE| (RFC-625) has been implemented
+following this process.  These \verb|DEFECTS|-like calibration products have
+their origins as human-curated text files in the appropriate
+\verb|obs_CAMERA_data| packages (and are thereby versioned through the
+packages' git repositories).  The text file must be in a structured
+format, to allow the header keywords to be read and assigned to a
+metadata PropertyList.
+
+For these products to be used, they must be converted into a FITS
+format that can be read by ingestCalibs.py.  This is performed by the
+\verb|pipe_tasks|/ingestCuratedCalibs.py script, which uses the metadata
+information and the calibration product's python class to read the
+text version and convert to a FITS file format that can be ingested by
+ingestCalibs.py.
+
+This updated calibration process removes the necessity that each
+\verb|obs_CAMERA_data| package implement one-off mapper methods to allow the
+product to be found.  However, only \verb|DEFECTS|, \verb|QE_CURVE|, and \verb|LINEARITY|
+follow this system, leaving the remaining products to be handled with
+camera-specific code and cameraGeom entries.  In addition to having
+diverse implementations, these handlers are often rigid, and cannot be
+versioned easily.
+
+\section{Transitioning to Gen3/What We Should Expect From Gen3}
+
+\subsection{A CalibrationBaseClass}
+The first step to a consistent calibration system is to ensure that
+all products in both generations use the \verb|DEFECTS|-like calibration
+product system.  For FITS image products, this requires future
+calibrations to be constructed with all of the header keyword values
+listed above.
+
+The remaining unconverted calibration types (\verb|CROSSTALK|, \verb|BFKERNEL|)
+should be converted to use \verb|DEFECTS|-like class, most likely by defining
+a CalibrationBaseClass, and making all of the calibration products
+subclasses of that.  This would enforce that all calibrations can be
+relied upon to provide getMetadata, setMetadata, readFits, writeFits,
```

Comment:
Why insist on FITS?  If YAML is efficient enough or something like Parquet is even more efficient, should those be options?

Reply:
What butler actually does with the `.put()` is down to the formatter we select so we can write text if we want. The fits writing was for compatibility with current behavior.

Reply:
_Mentioning_ FITS may be important, though, because if writing key-value pair metadata to FITS in a way that looks like reasonable to external code is a requirement or desire _when we use FITS_ (and I suspect it is), then FITS probably imposes some least-common-denominator limitations on what those key-value pairs can hold.

Reply:
We could use a table in FITS, not header keywords.  I think we should design what we need, independent of a 1980s standard inspired by FORTRAN 66.

Reply:
As above, I think FITS will be the default in-butler calibration format for speed purposes, with yaml used for exports.  I don't think we have any issues with the headers holding metadata now, and I've been adding more logic to the base class to help ensure that the common keywords are populated.

Reply:
It's not always YAML in text format.